src/stcal/ramp_fitting/likely_fit.py

```diff
@@ -0,0 +1,1353 @@
+                d2use, countrates = mask_jumps(
+                    diff[:, row], covar, readnoise_2d[row], gain_2d[row],
+                    threshold_oneomit=threshold_oneomit,
+                    threshold_twoomit=threshold_twoomit,
+                    diffs2use=d2use
+                )
+            else:
+                d2use, countrates = mask_jumps(
+                    diff[:, row], covar, readnoise_2d[row], gain_2d[row],
+                    diffs2use=d2use
+                )
+
+            # Set jump detection flags
+            jump_locs = d2use_copy ^ d2use
+            jump_locs[jump_locs > 0] = ramp_data.flags_jump_det
+            gdq[1:, row, :] |= jump_locs
+
+            alldiffs2use[:, row] = d2use
+
+            rateguess = countrates * (countrates > 0) + ramp_data.average_dark_current[row, :]
+            result = fit_ramps(
+                diff[:, row],
+                covar,
+                gain_2d[row],
+                readnoise_2d[row],
+                diffs2use=d2use,
+                countrateguess=rateguess,
+            )
+            integ_class.get_results(result, integ, row)
+
+        pdq = utils.dq_compress_sect(ramp_data, integ, gdq, pdq)
+        integ_class.dq[integ, :, :] = pdq
+
+        del gdq
+
+    integ_info = integ_class.prepare_info()
+    image_info = compute_image_info(integ_class, ramp_data)
+
+    return image_info, integ_info, opt_info
+
+
+def mask_jumps(
+    diffs,
+    Cov,
+    rnoise,
+    gain,
+    threshold_oneomit=20.25,
+    threshold_twoomit=23.8,
+    diffs2use=None,
+):
+
+    """
+    Function mask_jumps implements a likelihood-based, iterative jump
+    detection algorithm.
+
+    Parameters
+    ----------
+    diffs : ndarray
+        The group differences of the data array for a given integration and row
+        (ngroups-1, ncols).
+
+    Cov : Covar
+        The class that computes and contains the covariance matrix info.
+
+    rnoise : ndarray
+        The read noise (ncols,)
+
+    gain : ndarray
+        The gain (ncols,)
+
+    threshold_oneomit : float
+        Minimum chisq improvement to exclude a single resultant difference.
+        Default: 20.25.
+
+    threshold_twoomit : float
+        Minimum chisq improvement to exclude  two sequential resultant differences.
+        Default 23.8.
+
+    d2use : ndarray
+        A boolean array definined the segmented ramps for each pixel in a row.
+        (ngroups-1, ncols)
+
+    Returns
+    -------
+    d2use : ndarray
+        A boolean array definined the segmented ramps for each pixel in a row.
+        (ngroups-1, ncols)
+
+    countrates : ndarray
+        Count rate estimates used to estimate the covariance matrix.
+        Optional, default is None.
+
+    """
+    if Cov.pedestal:
+        raise ValueError(
+            "Cannot mask jumps with a Covar class that includes a pedestal fit."
+        )
+
+    # Force a copy of the input array for more efficient memory access.
+    loc_diff = diffs * 1
```

Comment:
This still needs addressing - why multiply by 1 here?

Reply:
In iterations after the first one the differences that need to be recomputed will be sparsely located in the full array of differences.  My intention was to force a copy to ensure that the ones that need to be recomputed will be contiguous in RAM.

Reply:
After a quick peek, I am not sure if this is actually helping; I will take a closer look.

Reply:
I remember now.  The array of differences is not contiguous in memory when operating on a row.  The first dimension is the number of reads and the second and third dimensions are the size of the array.  So, e.g., diffs[:, i] is not contiguous assuming that the last axis is the fast axis.  The intent of this line was to create an array that is contiguous in memory.  I don't know if it meaningfully improves performance.

Reply:
I have a hard time believing that memory access speed will win out over the performance hit of making a whole new numpy array - usual best practice with numpy is to avoid making a new array unless really necessary. But let's leave it for now, we can optimize performance later.